stratum/hal/lib/tdi/tdi_lut_manager.cc

```diff
@@ -0,0 +1,208 @@
+      }
+    }
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status TdiLutManager::DoWriteVlutEntry(
+    std::shared_ptr<TdiSdeInterface::SessionInterface> session,
+    uint32 bfrt_table_id, const ::p4::v1::Update::Type type,
+    const ::p4::v1::MvlutEntry& mvlut_entry) {
+  // Lock is already acquired by the caller
+  CHECK_RETURN_IF_FALSE(type != ::p4::v1::Update::UNSPECIFIED)
+      << "Invalid update type " << type;
+
+  ASSIGN_OR_RETURN(auto table_key,
+                   tdi_sde_interface_->CreateTableKey(bfrt_table_id));
+  RETURN_IF_ERROR(BuildTableKey(mvlut_entry, table_key.get()));
+
+  ASSIGN_OR_RETURN(
+      auto table_data,
+      tdi_sde_interface_->CreateTableData(bfrt_table_id, 0));
+
+  if (type == ::p4::v1::Update::INSERT || type == ::p4::v1::Update::MODIFY) {
+    for (const auto& param : mvlut_entry.param().params()) {
+      RETURN_IF_ERROR(table_data->SetParam(param.param_id(), param.value()));
+    }
+  }
+
+  switch (type) {
+    case ::p4::v1::Update::INSERT: {
+      RETURN_IF_ERROR(tdi_sde_interface_->InsertTableEntry(
+          device_, session, bfrt_table_id, table_key.get(),
+	  table_data.get()));
+      break;
+    }
+    case ::p4::v1::Update::MODIFY: {
+      RETURN_IF_ERROR(tdi_sde_interface_->ModifyTableEntry(
+          device_, session, bfrt_table_id, table_key.get(),
+	  table_data.get()));
+      break;
+    }
+    case ::p4::v1::Update::DELETE: {
+      RETURN_IF_ERROR(tdi_sde_interface_->DeleteTableEntry(
+          device_, session, bfrt_table_id, table_key.get()));
+      break;
+    }
+    default:
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported update type: " << type;
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status TdiLutManager::ReadTableEntry(
+    std::shared_ptr<TdiSdeInterface::SessionInterface> session,
+    const ::p4::v1::ExternEntry& entry,
+    WriterInterface<::p4::v1::ReadResponse>* writer) {
+  absl::ReaderMutexLock l(&lock_);
+
+  if (entry.extern_type_id() == kMvlutExactMatch ||
+      entry.extern_type_id() == kMvlutTernaryMatch) {
+      ::p4::v1::MvlutEntry mvlut_entry;
+      CHECK_RETURN_IF_FALSE(entry.entry().UnpackTo(&mvlut_entry))
+          << "Entry " << entry.ShortDebugString()
+          << " is not a mvlut entr.";
+      ASSIGN_OR_RETURN(uint32 bfrt_table_id,
+                   tdi_sde_interface_->GetTdiRtId(mvlut_entry.mvlut_id()));
+      return DoReadVlutEntry(session, bfrt_table_id,
+                             mvlut_entry, writer);
+  } else {
+      RETURN_ERROR(ERR_INVALID_PARAM)
+          << "Unsupported extern type " << entry.extern_type_id() << ".";
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status TdiLutManager::DoReadVlutEntry(
+    std::shared_ptr<TdiSdeInterface::SessionInterface> session,
+    uint32 table_id,
+    const ::p4::v1::MvlutEntry& mvlut_entry,
+    WriterInterface<::p4::v1::ReadResponse>* writer) {
+
+    return MAKE_ERROR(ERR_INTERNAL) << "READ VLUT ENTRY IS YET TO BE SUPPORTED";
```

Comment:
"Not yet supported". "Not yet to be" is future conditional (I believe). And the error text should not be upper-case.
(I CAN HEAR YOU, THERE'S NO NEED TO SHOUT!)